zapgrpc/zapgrpc.go

```diff
@@ -36,65 +39,129 @@ func (f optionFunc) apply(log *Logger) {
-}
-
-// Fatal implements grpclog.Logger.
-func (l *Logger) Fatal(args ...interface{}) {
-	l.fatal(l.log, args...)
-}
-
-// Fatalf implements grpclog.Logger.
-func (l *Logger) Fatalf(format string, args ...interface{}) {
-	l.fatalf(l.log, format, args...)
-}
-
-// Fatalln implements grpclog.Logger.
-func (l *Logger) Fatalln(args ...interface{}) {
-	l.fatal(l.log, args...)
+	delegate     *zap.SugaredLogger
+	printToDebug bool
+	fatalToWarn  bool
 }
 
 // Print implements grpclog.Logger.
+// Deprecated: use Info().
 func (l *Logger) Print(args ...interface{}) {
-	l.print(l.log, args...)
+	if l.printToDebug {
+		l.delegate.Debug(args...)
+	} else {
+		l.delegate.Info(args...)
+	}
 }
 
 // Printf implements grpclog.Logger.
+// Deprecated: use Infof().
 func (l *Logger) Printf(format string, args ...interface{}) {
-	l.printf(l.log, format, args...)
+	if l.printToDebug {
+		l.delegate.Debugf(format, args...)
+	} else {
+		l.delegate.Infof(format, args...)
+	}
 }
 
 // Println implements grpclog.Logger.
+// Deprecated: use Info().
 func (l *Logger) Println(args ...interface{}) {
-	l.print(l.log, args...)
+	l.Print(args...)
+}
+
+// Info implements grpclog.LoggerV2.
+func (l *Logger) Info(args ...interface{}) {
+	l.delegate.Info(args...)
+}
+
+// Infoln implements grpclog.LoggerV2.
+func (l *Logger) Infoln(args ...interface{}) {
+	l.delegate.Info(args...)
+}
+
+// Infof implements grpclog.LoggerV2.
+func (l *Logger) Infof(format string, args ...interface{}) {
+	l.delegate.Infof(format, args...)
+}
+
+// Warning implements grpclog.LoggerV2.
+func (l *Logger) Warning(args ...interface{}) {
+	l.delegate.Warn(args...)
+}
+
+// Warningln implements grpclog.LoggerV2.
+func (l *Logger) Warningln(args ...interface{}) {
+	l.delegate.Warn(args...)
+}
+
+// Warningf implements grpclog.LoggerV2.
+func (l *Logger) Warningf(format string, args ...interface{}) {
+	l.delegate.Warnf(format, args...)
+}
+
+// Error implements grpclog.LoggerV2.
+func (l *Logger) Error(args ...interface{}) {
+	l.delegate.Error(args...)
+}
+
+// Errorln implements grpclog.LoggerV2.
+func (l *Logger) Errorln(args ...interface{}) {
+	l.delegate.Error(args...)
+}
+
+// Errorf implements grpclog.LoggerV2.
+func (l *Logger) Errorf(format string, args ...interface{}) {
+	l.delegate.Errorf(format, args...)
+}
+
+// Fatal implements grpclog.LoggerV2.
+func (l *Logger) Fatal(args ...interface{}) {
+	if l.fatalToWarn {
+		l.delegate.Warn(args...)
+	} else {
+		l.delegate.Fatal(args...)
+	}
+}
+
+// Fatalln implements grpclog.LoggerV2.
+func (l *Logger) Fatalln(args ...interface{}) {
+	l.Fatal(args...)
+}
+
+// Fatalf implements grpclog.LoggerV2.
+func (l *Logger) Fatalf(format string, args ...interface{}) {
+	if l.fatalToWarn {
+		l.delegate.Warnf(format, args...)
+	} else {
+		l.delegate.Fatalf(format, args...)
+	}
+}
+
+// V implements grpclog.LoggerV2.
+func (l *Logger) V(level int) bool {
+	return l.delegate.Desugar().Core().Enabled(zapcore.Level(level))
```

Comment:
I'm not sure the gRPC levels are directly convertible to zap levels,

Based on the code here,
https://github.com/grpc/grpc-go/blob/v1.35.0/grpclog/loggerv2.go#L79

It looks like gRPC maps:
 - 0=info
 - 1=warn
 - 2=error
 - 3=fatal

zap is similar, but slightly different
https://pkg.go.dev/go.uber.org/zap@v1.16.0/zapcore#Level
 - 0=info
 - 1=warn
 - 2=error
 - 3=dpanic

We should convert gRPC's fatal (3) to a 5 to map to zap.

Can we add tests to validate the gRPC / zap levels map correctly as well?

Reply:
@prashantv Done, PTAL.